### PR TITLE
8224267: JOptionPane message string with 5000+ newlines produces StackOverflowError

### PR DESCRIPTION
**Pull request by prsadhuk**
BasicOptionPaneUI.addMessageComponents() uses recursion to split message strings at newlines, generating a StackOverflowError when the message string contains an unusually large number of newlines.
Fixed by catching StackOverflow and ignoring so that application is not stuck.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8224267](https://bugs.openjdk.org/browse/JDK-8224267): JOptionPane message string with 5000+ newlines produces StackOverflowError

### Reviewers
 * [Tejesh R](https://openjdk.org/census#tr) (@TejeshR13 - Author) ⚠️ Review applies to [9376b0ff](https://git.openjdk.org/jdk/pull/9388/files/9376b0ff26216c36cf9464725e5e139403cdd93e)
 * [Alexey Ivanov](https://openjdk.org/census#aivanov) (@aivanov-jdk - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9388/head:pull/9388` \
`$ git checkout pull/9388`

Update a local copy of the PR: \
`$ git checkout pull/9388` \
`$ git pull https://git.openjdk.org/jdk pull/9388/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9388`

View PR using the GUI difftool: \
`$ git pr show -t 9388`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9388.diff">https://git.openjdk.org/jdk/pull/9388.diff</a>

</details>

### CODE REVIEW

**Comment by prsadhuk** [comment]
PR

**Review by TejeshR13 — APPROVED** [review]

**Review by aivanov-jdk — COMMENTED** [review]
Maybe we should rather limit the number of recursive calls? You caught `StackOverflowError`, at this stage there's not much space left on the stack. It may cause another `StackOverflowError`.

**Comment by aivanov-jdk** [comment]
> > Maybe we should rather limit the number of recursive calls? You caught `StackOverflowError`, at this stage there's not much space left on the stack. It may cause another `StackOverflowError`.
> 
> I thought that but then thought limiting will not be foolproof as stack size can be modified by -Xss.

If the limit is low, let's say 20-30 lines, it shouldn't be a problem.

I'm afraid a message with more than 200 lines won't fit on the screen anyway.

**Comment by prsadhuk** [comment]
> > > Maybe we should rather limit the number of recursive calls? You caught `StackOverflowError`, at this stage there's not much space left on the stack. It may cause another `StackOverflowError`.
> > 
> > 
> > I thought that but then thought limiting will not be foolproof as stack size can be modified by -Xss.
> 
> If the limit is low, let's say 20-30 lines, it shouldn't be a problem.
> 
> I'm afraid a message with more than 200 lines won't fit on the screen anyway.

OK. Seems logical..I guess it could be much less lines that fit in screen but anyways, to be safe I restrict to this.

File: src/java.desktop/share/classes/javax/swing/plaf/basic/BasicOptionPaneUI.java [hunk 1]
```diff
@@ -455,32 +456,35 @@ protected void addMessageComponents(Container container,
...
+                    };
+                    breakPanel.setName("OptionPane.break");
+                    addMessageComponents(container, cons, breakPanel, maxll,
+                                         true);
+                } else {
+                    addMessageComponents(container, cons, s.substring(0, nl),
+                                      maxll, false);
+                }
+                // Prevent recursion of more than
+                // 200 successive newlines in a message
```
Comment (aivanov-jdk) [inline src/java.desktop/share/classes/javax/swing/plaf/basic/BasicOptionPaneUI.java:480]: Not necessarily ‘successive’, but I'm nitpicking.

**Review by aivanov-jdk — APPROVED** [review]

**Comment by prsadhuk** [comment]
I guess if it's not successive, it will not exercise this condition so it will not recurse as much
```
if (nl >= 0) {
                // break up newlines
                if (nl == 0) {
```

**Comment by aivanov-jdk** [comment]
> I guess if it's not successive, it will not exercise this condition so it will not recurse as much
> 
> ```
> if (nl >= 0) {
>                 // break up newlines
>                 if (nl == 0) {
> ```

In that sense? I thought about successive in this case like `\n\n\n` but it applies to *all the new lines* in a one message.

**Comment by aivanov-jdk** [comment]
Hm… there could be problem: **the counter isn't reset** when the function is called again. It's reset only when the limit is reached. Or do I miss another case where the counter is reset?

The easiest way could be to introduce a helper function which is called from the main function that breaks the message and to which the recursion counter is passed as a parameter.

**Comment by prsadhuk** [comment]
> Hm… there could be problem: **the counter isn't reset** when the function is called again. It's reset only when the limit is reached. Or do I miss another case where the counter is reset?
> 
> The easiest way could be to introduce a helper function which is called from the main function that breaks the message and to which the recursion counter is passed as a parameter.

I dont think it's a problem with other cases as in those "msg" is non String and so wont recurse as much as successive newline case in this condition, so I have made it to reset in this.

**Comment by aivanov-jdk** [comment]
> > Hm… there could be problem: **the counter isn't reset** when the function is called again. It's reset only when the limit is reached. Or do I miss another case where the counter is reset?
> >
> > The easiest way could be to introduce a helper function which is called from the main function that breaks the message and to which the recursion counter is passed as a parameter.
> 
> I dont think it's a problem with other cases as in those "msg" is non String and so wont recurse as much as successive newline case in this condition, so I have made it to reset in this.

If the message contains new lines, let's say 20 new lines each. Then I call `JOptionPane.showMessageDialog` continuously with the same message. Since the counter is not reset between the calls, the 10th call would stop breaking the message at the first new line in the message, which doesn't look correct.

I think we have just introduced a new bug which needs resolving.

As another thought, shall we indicate the original message is truncated? Add `...` (ellipsis) to the previous line of text?

**Comment by mrserb** [comment]
> shall we indicate the original message is truncated? Add ... (ellipsis) to the previous line of text?

I think this is a great idea.

**Comment by j3graham** [comment]
An alternative approach would be to avoid recursion when splitting the string.

Instead, create a new private method that does the string processing for strings without newlines. Then split the string on newlines, and call that new method for each one. “String.lines” might be a close enough match to do the splitting - one difference would be that it always splits on \r, \n and \r\n instead of looking at the system newline value.

**Pull request closed without merging**